src/pubsub.js

```diff
@@ -0,0 +1,573 @@
+/* eslint-env mocha */
+/* eslint max-nested-callbacks: ['error', 8] */
+'use strict'
+
+const expect = require('chai').expect
+const series = require('async/series')
+const waterfall = require('async/waterfall')
+const parallel = require('async/parallel')
+const whilst = require('async/whilst')
+const each = require('async/each')
+
+function waitForPeers (ipfs, topic, peersToWait, callback) {
+  const i = setInterval(() => {
+    ipfs.pubsub.peers(topic, (err, peers) => {
+      if (err) {
+        return callback(err)
+      }
+
+      const missingPeers = peersToWait
+            .map((e) => peers.includes(e))
+            .filter((e) => !e)
+
+      if (missingPeers.length === 0) {
+        clearInterval(i)
+        callback()
+      }
+    })
+  }, 500)
+}
+
+function spawnWithId (factory, callback) {
+  waterfall([
+    (cb) => factory.spawnNode(cb),
+    (node, cb) => node.id((err, res) => {
+      if (err) {
+        return cb(err)
+      }
+      node.peerId = res
+      cb(null, node)
+    })
+  ], callback)
+}
+
+function makeCheck (n, done) {
+  let i = 0
+  return (err) => {
+    if (err) {
+      return done(err)
+    }
+
+    if (++i === n) {
+      done()
+    }
+  }
+}
+
+module.exports = (common) => {
+  describe('.pubsub', () => {
+    const topic = 'pubsub-tests'
+
+    describe('callback API', () => {
+      let ipfs1
+      let ipfs2
+      let ipfs3
+
+      before((done) => {
+        common.setup((err, factory) => {
+          if (err) {
+            return done(err)
+          }
+
+          series([
+            (cb) => spawnWithId(factory, cb),
+            (cb) => spawnWithId(factory, cb),
+            (cb) => spawnWithId(factory, cb)
+          ], (err, nodes) => {
+            if (err) {
+              return done(err)
+            }
+
+            ipfs1 = nodes[0]
+            ipfs2 = nodes[1]
+            ipfs3 = nodes[2]
+            done()
+          })
+        })
+      })
+
+      after((done) => {
+        common.teardown(done)
+      })
+
+      describe('single node', () => {
+        describe('.publish', () => {
+          it('message from string', (done) => {
+            ipfs1.pubsub.publish(topic, 'hello friend', done)
+          })
```

Comment:
I agree @diasdavid that this is something we should decide now as it has bitten us already in a bad way. What do we do in other function that accept similar data a param, like object.put? It might be a good idea to follow similar convention. If it's string+buffer, then what David proposes makes sense (always convert to buffer and make it as explicit to the user as possible (docs + erroring on anything but string or buffer). If it's only a buffer, it would also make sense to me (input buffer, receive buffer). This is how I tried to approach it in js-ipfs-api: everything we receive is a string, so inputs should be strings (from user's perspective). Obviously we would need to change js-ipfs-api to convert all strings to buffers, and probably in a lot of use cases the user then converts the buffer back to a string, but I suppose it's a cost we could live with (and on a positive side, js-ipfs would be tiny bit faster because it only needs to use buffers).

So... I think I like the idea of accepting buffers-only the most, but it wouldn't be too bad if we accepted strings too.